Creating-Objects.md

```diff
@@ -1,12 +1,105 @@
+* Provide a `normalize` class method that accepts a raw Python object. The method should first validate the raw object to check whether it is malformed. If the validation checks pass, then the method should return a normalized form of the object. Note that in this context, "normalized" means a primitive Python type or multiple primitive types combined with lists and/or dicts.
+
+  * The `BaseObject` class provides a `normalize` class method that can handle normalization for you so long as your object class provides a `get_schema` class method that returns a schema describing the object. `BaseObject.normalize` passes your schema to `schema_utils.normalize_against_schema`, so make sure that `normalize_against_schema` won't try to normalize your object to your schema using your object class. Otherwise, you'll get an endless loop between `normalize_against_schema` and `BaseObject.normalize`.
+
+### Angular components
+
+Every object is accompanied by a component defined in `extensions/objects/templates` and imported in `extensions/objects/object-components.module.ts`. Each component provides the HTML and frontend code needed to create a form for the user to provide input in the form of the object. These components rely heavily on [schema-based forms](https://github.com/oppia/oppia/wiki/Schemas#schema-based-forms). For example, the `int-editor` component's HTML is just this:
+
+```html
+<schema-based-editor [schema]="getSchema.bind(this)"
+                     [localValue]="value"
+                     (localValueChange)="updateValue($event)">
+</schema-based-editor>
+```
+
+The `getSchema` function is defined in `int-editor.component.ts` and returns this schema:
+
+```json
+{
+  type: "int",
+  validators: [{
+    id: "is_integer"
+  }]
+}
+```
+
+The `schema-based-editor` component already knows how to construct a form for this schema.
+
+## Create a new object
+
+Let's suppose you're creating a new object called `MyObject` (in practice you should use a more descriptive name). You'll need to follow these steps:
+
+1. Create a new [object class](#python-classes) named `MyObject`.
+2. Add tests for your class in `extensions/objects/models/objects_test.py`. You should add normalization tests to the `ObjectNormalizationUnitTests` class. You can also create a new test class to hold additional test cases.
+3. Add a new [Angular component](#angular-components) to accept input in the form of your object. Your component files should be named `my-object-editor.component.*`, and your component class should be named `MyObjectEditorComponent`.
+4. Import your component in `extensions/objects/object-components.module.ts` and add it to the module there.
+
+## Using objects
+
+You are unlikely to use objects directly. Instead, you will reference them as you [[create interactions|Creating-Interactions]], [[define rules|Creating-Rules]], or otherwise use [[schemas|Schemas]].
```

Comment:
Not really sure what the alternative of "use objects directly" means -- aren't the latter cases examples of "using objects directly"?

Reply:
I removed that bit